### ISSUE TITLE
specifing a GPU index

### ISSUE THREAD

**Issue by mirkobronzi** [tensorflow/tensor2tensor#200]
Hi, I was wondering what is the 'cleanest' way to specify a GPU index to run on.
(e.g., if you have a machine with 8 GPUs, and you have been allocated GPU number 5)

I am currently able to do that by hacking the config_proto, but this means code duplication..
is there an option to do so?

Like, an option for running on one GPU, and at the same time specify which GPU id to use (like CUDA_VISIBLE_DEVICE).

At the moment, if I do not hack the config_proto, my experiments will just run on CPU (even if I use --worker_gpu or --ps_gpu)

**Comment by martinpopel** [comment]
I am using `CUDA_VISIBLE_DEVICE` and it works fine with Tensor2tensor.

**Comment by mirkobronzi** [comment]
thanks for the info.
seems not to work for me, but I assume some mistake on my part then..

**Comment by mirkobronzi** [comment]
Well, after double checking (and logging env variables), I think I am setting CUDA_VISIBLE_DEVICE correctly..

looking at the code, does anyone know where the CUDA_VISIBLE_DEVICE is (supposed to be) handled?

Like, if I look at the code where the config_proto is created, i.e.,
https://github.com/tensorflow/tensor2tensor/blob/master/tensor2tensor/utils/trainer_utils.py#L327
I see that no 'visible_device_list' is set.

So that I end up with:
gpu_options {
  per_process_gpu_memory_fraction: 1.0
}

instead of (this is the one working - when I hack the config_proto):
gpu_options {
  per_process_gpu_memory_fraction: 1.0
  visible_device_list: "14"
}

Any insight?

Thanks a lot,
Mirko

**Comment by rsepassi** [comment]
The environment variable to be set is `CUDA_VISIBLE_DEVICES`. This is not dealt with by Tensor2Tensor or TensorFlow but is specific to CUDA. With the env var set, the process only sees those devices.

**Comment by rsepassi** [comment]
Closing for now as it does not seem that this is a bug. Please do reopen if you find differently.

**Comment by martinpopel** [comment]
I just confirm, I am using `CUDA_VISIBLE_DEVICES`, not `CUDA_VISIBLE_DEVICE`.

**Comment by mirkobronzi** [comment]
same here - using the correct one (the typo is just in this thread - sorry for that) - still, not working.

Just to be clear - my problem is that if I do not hack the code directly, tensor2tensor runs on CPU (not on the wrong GPU - it just falls back on CPU).

So, not sure there are some additional GPU option required to use a GPU (I tried the various option in t2t-train, but no luck) .

As I showed above, the problem seems to be that the config_proto is simply not specifying any visible device GPU id, i.e.,
```
gpu_options {
per_process_gpu_memory_fraction: 1.0
}
```

this is way it could help to know how tensor2tensor decides to run on GPU or CPU.
(is it enough to use a GPU compiled version of tensorflow? it would seem not from my experiment, given I am using that one)

Thanks again,
Mirko
